Final/RFC00XX-Custom-Remote-Connections.md

```diff
@@ -0,0 +1,281 @@
+
+- Create a text writer wrapper object, that is used to write protocol messages to the target input data stream.
+
+- Start a listening thread for protocol messages from the target output data stream.
+
+### Data Streams
+
+All connection data streams used in the transport must be derived from .NET TextReader or TextWriter classes.
+Connections typically have two data streams, one for writing messages to the target, and one for reading messages from the target.
+However, some connections have a third error stream that require a special reader thread.
+An example would be a Process connection that uses stdOut for reading from the target, stdIn for writing to the client, and stdError for reading connection error messages.
+The stdError messages are not part of the remoting protocol and need to be handled separately.
+
+### Creating a connection
+
+The PowerShell module implementing the new remote connection should supply all necessary cmdlets for creating and managing the connection.
+In particular it should have a cmdlet that returns a `PSSession` object.
+
+```powershell
+PS C:\> Import-Module -Name MyCustomConnection
+PS C:\> New-MyCustomSession -ComputerName remoteVM1 -Cred $myCredentials
+
+Id Name            Transport ComputerName    ComputerType    State         ConfigurationName     Availability
+ -- ----            --------- ------------    ------------    -----         -----------------     ------------
+  1 Runspace1       MyCustom  remoteVM1       RemoteMachine   Opened        DefaultShell             Available
+```
+
+However, if the new connection info class is made public accessible, the PowerShell remoting API can also be used to create a connection instance.
+
+```powershell
+PS C:\> Import-Module -Name MyCustomConnection
+PS C:\> $connectionInfo = [MyCustomConnection.CustomConnectionInfo]::new($computerName, $creds)
+PS C:\> $runspace = [runspacefactory]::CreateRunspace($connectionInfo)
+PS C:\> $runspace.Open()
+PS C:\> $runspace
+
+ Id Name            ComputerName    Type          State         Availability
+ -- ----            ------------    ----          -----         ------------
+  2 Runspace2       remote-1        Remote        Opened        Available
+```
+
+### Endpoints
+
+A remote connection must include a PowerShell session on the target machine, that handles remoting protocol messages from the client.
+This is called the remoting endpoint.
+An endpoint typically involves a service that listens for client connection requests, and securely creates the PowerShell session under the correct account and with correct permissions.
+For example, both WSMan and SSH have services that do this.
+Endpoints such as these are outside the scope of this feature proposal, and it is assumed that any custom connection scheme will include some sort of endpoint support.  
+
+PowerShell currently supports a 'server' running mode, where remoting protocol messages are channeled through the process stdIn/stdOut data streams.
```

Comment:
Should the named pipe server also be mentioned, i.e. not the `-s` option but the `PSHost.` named pipe that pwsh creates for stuff like `Enter-PSHostProcess`/`-CustomPipeName`?

Reply:
I don't see any need to include other server modes in this proposal.  To keep things simple, I just included proc/stdIO mode which I feel most implementers would use.  Another endpoint option is to write a PowerShell WinRM plugin host since, since that is actually the only other way to create a PowerShell remoting endpoint instance.  But it is very complex and doesn't provide much value in my opinion.

Reply:
Just my two cents: the way the OpenSSH client and server integration is currently done (subprocess + standard input/output streams) fits my needs perfectly except for the fact that there is no simple way to override the executable to launch (and its parameters). If this is exposed in some generic way I'd be happy, but I'm also happy if it's just exposed in the SSH remoting transport. I like it because it requires no modification to PowerShell whatsoever and no special classes, registration, etc.

Reply:
While that would be nice, it's not much more of a step removed to enable a class implementation and just ask the user to install a module with your custom transport. It's already been sadly discussed to death that modifying that existing provider isn't going to fly, however this solution allows it to be reimplemented to do what you want at least.